### PR TITLE
Add gRPC error codes

### PR DESCRIPTION
**Pull request by shgandhi**
This commit updates TRC/TRS/CC/CS to support gRPC error codes of OUT_OF_RANGE, NOT_FOUND and INTERNAL. Please see individual commits for more detail.
The unit tests will be added in a follow-up PR.

### CODE REVIEW

**Review by robem — CHANGES_REQUESTED** [review]

File: client/thin-replica-client/include/client/thin-replica-client/thin_replica_client.hpp [hunk 1]
```diff
@@ -423,6 +423,24 @@ class ThinReplicaClient final {
...
+    UpdateNotFound() : std::runtime_error("requested update does not exist yet"){};
+  };
+
+  // A new subscription may request an out of range update
+  class OutOfRangeSubscriptionRequest : public std::runtime_error {
+   public:
+    OutOfRangeSubscriptionRequest() : std::runtime_error("out of range subscription request"){};
+  };
+
+  // An internal error may occur, for example when max_agreeing
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply client/thin-replica-client/include/client/thin-replica-client/thin_replica_client.hpp]: Thanks for catching, corrected.

**Comment by shgandhi** [comment]
Thanks for the review @robem. Updated the PR acc. to our discussion, and your comments.

File: client/concordclient/src/concord_client.cpp [hunk 2]
```diff
@@ -147,9 +147,25 @@ void ConcordClient::subscribe(const SubscribeRequest& sub_req,
   if (std::holds_alternative<EventGroupRequest>(sub_req.request)) {
     ::client::thin_replica_client::SubscribeRequest trc_request;
     trc_request.event_group_id = std::get<EventGroupRequest>(sub_req.request).event_group_id;
-    trc_->Subscribe(trc_request);
+    try {
+      trc_->Subscribe(trc_request);
+    } catch (ThinReplicaClient::UpdateNotFound& e) {
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply client/concordclient/src/concord_client.cpp]: Yes, I agree. I have moved the exceptions to a separate header file `concord_client_exceptions.hpp`.

File: kvbc/test/kvbc_app_filter/kvbc_filter_test.cpp [hunk 3]
```diff
@@ -578,12 +576,42 @@ TEST(kvbc_filter_test, kvbfilter_event_group_out_of_range_eg) {
...
   EXPECT_THROW(kvb_filter.readEventGroupRange(eg_id, queue_out, stop_exec);, InvalidEventGroupRange);
 }
 
+TEST(kvbc_filter_test, kvbfilter_get_oldest_tag_specific_eg_id_pvt) {
+  FakeStorage storage;
+  std::string client_id("trid_1");
+  auto kvb_filter = KvbAppFilter(&storage, client_id);
+  size_t num_event_groups_to_fill = 5;
+  storage.fillWithEventGroupData(num_event_groups_to_fill, client_id);
+  spsc_queue<KvbFilteredEventGroupUpdate> queue_out{num_event_groups_to_fill};
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply kvbc/test/kvbc_app_filter/kvbc_filter_test.cpp]: Thanks, updated.

File: kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp [hunk 4]
```diff
@@ -314,6 +314,8 @@ uint64_t KvbAppFilter::oldestTagSpecificPublicEventGroupId() {
   uint64_t public_oldest = getValueFromLatestTable(kPublicEgIdKeyOldest);
   uint64_t private_oldest = getValueFromLatestTable(client_id_ + "_oldest");
   if (!public_oldest && !private_oldest) return 0;
+  if (!public_oldest) return private_oldest;
+  if (!private_oldest) return public_oldest;
   return public_oldest + private_oldest - 1;
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp:322]: Yes, added the comment.

File: thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp [hunk 5]
```diff
@@ -717,6 +791,13 @@ class ThinReplicaImpl {
                    ServerWriterT* stream,
                    std::shared_ptr<kvbc::KvbAppFilter>& kvb_filter,
                    bool event_group_enabled = false) {
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp]: Thanks, updated.

**Comment by shgandhi** [comment]
Thanks for the review @robem, updated the MR as suggested.

File: client/concordclient/src/concord_client.cpp [hunk 6]
```diff
@@ -147,9 +147,25 @@ void ConcordClient::subscribe(const SubscribeRequest& sub_req,
...
-    trc_->Subscribe(trc_request);
+    try {
+      trc_->Subscribe(trc_request);
+    } catch (UpdateNotFound& e) {
+      throw;
+    } catch (OutOfRangeSubscriptionRequest& e) {
+      throw;
+    } catch (InternalError& e) {
+      throw;
+    }
```
Comment (robem) [inline client/concordclient/src/concord_client.cpp:158]: Given that you don't do anything we the exceptions, you can simply get rid of the try-catch block.
Same below.

Comment (shgandhi) [reply client/concordclient/src/concord_client.cpp:158]: We throw the same exception, so that it can be caught by the event_service.cpp.

Comment (robem) [reply client/concordclient/src/concord_client.cpp:158]: The exceptions are thrown already, you don't have to catch and throw them again.

**Pull request merged**